backend/pong/game_logic.py

```diff
@@ -0,0 +1,113 @@
+# pong/game_logic.py
+from .constants import *
+import math
+
+
+class PongGame:
+    def __init__(self):
+        # Initialize game state
+        self.player1_y = PLAYGROUND_HEIGHT // 2  # Initial position for player 1 (midpoint)
+        self.player2_x = PLAYGROUND_WIDTH - WALL_MARGIN
+        self.player1_x = WALL_MARGIN
+        self.player2_y = PLAYGROUND_HEIGHT // 2  # Initial position for player 2 (midpoint)
+        self.ball_x = PLAYGROUND_WIDTH - WALL_MARGIN - PLAYER_WIDTH   # Initial position for ball (x-coordinate)
+        self.ball_y = PLAYGROUND_HEIGHT // 2     # Initial position for ball (y-coordinate)
+        self.ball_angle = 0  # Initial angle of the ball's movement
+        self.game_started = False  # Indicates if the game has started
+        self.game_over = False  # Indicates if the game is over
+        self.winner = None  # Player who won
+        self.loser = None  # Player who lost
+
+    def update_positions(self, key_press):
+        self.update_player_positions(key_press)
+        self.update_ball_position()
+
+    def update_player_positions(self, key_press):
+        # Update player positions based on key press
+        if not self.game_started and key_press in ('o', 'l'):
+            # If player 2 moves, start the game
+            self.kick_start_game(key_press)
+        elif key_press == 'w':
+            self.player1_y -= PLAYER_MOVEMENT_UNIT  # Move player 1 up
+        elif key_press == 's':
+            self.player1_y += PLAYER_MOVEMENT_UNIT  # Move player 1 down
+        elif key_press == 'o':
+            self.player2_y -= PLAYER_MOVEMENT_UNIT  # Move player 2 up
+        elif key_press == 'l':
+            self.player2_y += PLAYER_MOVEMENT_UNIT  # Move player 2 down
+        
+        # Ensure player positions stay within the bounds of the play area
+        self.player1_y = max(min(self.player1_y, PLAYGROUND_HEIGHT - PLAYER_HEIGHT), 0)  # Bound player 1 between 0 and (playground height - player height)
+        self.player2_y = max(min(self.player2_y, PLAYGROUND_HEIGHT - PLAYER_HEIGHT), 0)  # Bound player 2 between 0 and (playground height - player height)
+    def kick_start_game(self, key_press):
+        # Kick start the game based on player 2's movement
+        if key_press == 'o':
+            # Player 2 moves upward, set ball angle to 120 degrees
+            self.ball_angle = 120
+        elif key_press == 'l':
+            # Player 2 moves downward, set ball angle to 240 degrees
+            self.ball_angle = 240
+        
+        self.game_started = True
+    
+    # Update ball position based on angle
+    def update_ball_position(self):
+        if self.game_started:
+            self.ball_x += BALL_SPEED * math.cos(math.radians(self.ball_angle))
+            self.ball_y += BALL_SPEED * math.sin(math.radians(self.ball_angle))
+            self.check_collisions()  # Check for collisions
+
+    def check_collisions(self):
```

Comment:
We might need some kind of tolerance value here to avoid the dudes actually intersecting, which is going to be a little ugly